### PR TITLE
feat: passcode table

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

check LOG-1181 to see details.

@logto-io/eng

### CODE REVIEW

File: packages/schemas/tables/passcodes.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+create type passcode_type as enum ('SignIn', 'Register', 'ForgotPassword');
+
+create table passcodes (
+  id varchar(128) not null,
+  session_id varchar(128) not null,
+  phone varchar(32),
+  email varchar(128),
```
Comment (xiaoyijun) [inline packages/schemas/tables/passcodes.sql:7]: Maybe a type (`email`| `phone`) and a `target` field would be better?
Is it possible to add new ways for users to receive a passcode? e.g. a code generated by an app, like Microsoft Authenticator

Comment (gao-sun) [reply packages/schemas/tables/passcodes.sql:7]: 1. good point, we may have other ways (like push)
2. MS Authenticator is time-based OTP, which does not need to generate passcode from Logto

File: packages/schemas/tables/passcodes.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  type passcode_type not null,
+  code varchar(6) not null,
+  used boolean not null default TRUE,
+  try_count int2 not null default 0,
+  created_at timestamptz not null default(now()),
+  primary key (id)
```
Comment (gao-sun) [inline packages/schemas/tables/passcodes.sql:13]: also foreign key?

File: packages/schemas/tables/passcodes.sql (hunk 1 above)
Comment (wangsijie) [reply packages/schemas/tables/passcodes.sql:7]: a session may have both `phone` and `email` passcode, any ideas? @logto-io/eng

File: packages/schemas/tables/passcodes.sql (hunk 2 above)
Comment (wangsijie) [reply packages/schemas/tables/passcodes.sql:13]: `session` doesn't have its own table, is inside `oidc_model_instances`, so maybe don't need a foreign key.

File: packages/schemas/tables/passcodes.sql (hunk 1 above)
Comment (simeng-li) [reply packages/schemas/tables/passcodes.sql:7]: besides would phone || email be required? At least one must provided.

File: packages/schemas/tables/passcodes.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+create type passcode_type as enum ('SignIn', 'Register', 'ForgotPassword');
+
+create table passcodes (
+  id varchar(128) not null,
+  session_id varchar(128) not null,
+  phone varchar(32),
+  email varchar(128),
+  type passcode_type not null,
+  code varchar(6) not null,
+  used boolean not null default TRUE,
```
Comment (simeng-li) [inline packages/schemas/tables/passcodes.sql:10]: nit: for OIDC tokens they use `consumed`.  just FYI.

File: packages/schemas/tables/passcodes.sql (hunk 2 above)
Comment (simeng-li) [reply packages/schemas/tables/passcodes.sql:13]: should session_id be a foreign key?

File: packages/schemas/tables/passcodes.sql (hunk 3 above)
Comment (darcyYe) [inline packages/schemas/tables/passcodes.sql:10]: wondering why the default value of `used` is TRUE not FALSE?

Comment (wangsijie) [reply packages/schemas/tables/passcodes.sql:10]: my fault

File: packages/schemas/tables/passcodes.sql (hunk 1 above)
Comment (wangsijie) [reply packages/schemas/tables/passcodes.sql:7]: Yes, but is hard to add this constraint in DB

**Review by darcyYe — APPROVED** [review]
LGTM

File: packages/schemas/tables/passcodes.sql (hunk 1 above)
Comment (simeng-li) [reply packages/schemas/tables/passcodes.sql:7]: I was thinking a field like `indicator not null` could be either phone and email or any. and another field indictor_type: enum. phone , email sth like that.

Comment (simeng-li) [reply packages/schemas/tables/passcodes.sql:7]: feel free to resolve this one, if we haven;t come up with a better solution.

**Pull request merged**